specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

```diff
@@ -175,6 +182,107 @@ In this context, a component is any separate software process.
+- provide a configuration step that lets users call an api and configure the
+  format, e.g.
+```
+logs.format().use()
+  .timestamp()
+  .job_name()
+  .level()
+```
+  - not sure how it will work on an environment level, maybe we can provide it
+    on the CLI / in the helm config
+- it should be extendable, e.g. if we want to add some new classification to the
+  format (parent job, team name), it should be easy to do in a modular fasion
+
+### Log Less
+
+#### Logs not at the apprpopriate level
+
+`[Local][Cloud]` We have to go through every log statement in every VDK plugin and make sure it's
+at the appropriate level. Judging by user feedback, we have a lot of INFO
+logging that should be at the DEBUG level.
+
+`[Local][Cloud]` There should be an easy way to for the user configure the datajob log level
+dynamically.
+
+#### Multi-layered logging
+
+`[Local]` Logs on failure should point to the problem and provide the minimum
+amount of troubleshooting information to get started. Any other information,
+e.g. full stack trace, should be output to a temp file.
+
+```
+process_twitter_data Step 10_ingest_data.py Line: 38
+File with path "./some_file.txt does not exist
+Full error log /tmp/30618c1b-677b-4f96-86a3-dda26011b3d8-1690469444-20a99/error.log
+```
+`[Cloud]` Full error log should be output to stdout
+
+### Don't Repeat Yourself (DRY)
+
+`[Local][Cloud]` Remove any repeating info in success logs. Success logs should
+show which step has started or ended and the success status at the end. Note
+that locally, this could be entirely replaced by progress bars, but users should
+be able to choose between one or the other.
+
+### Clean Error Handling
+
+`[Local][Cloud]` Do away with the log-and-throw and log-and-rethrow patterns.
+Errors should be passed up the call stack to the original caller. Errors should
+be caught and wrapped or replaced by other errors only if we're adding more
+information to the existing error.
+
+To preserve information, we can use an error context, parse it and dump it to
+stdout or a log file, depending on configuration. Recoverable errors should not
+be passed up the call stack, but just logged at the `WARN` level. Errors we
+can't recover from should be added to the context and re-thrown.
+
+`[Local][Cloud]` Error classification is good, because it eliminates firction
+between users and support teams. The current error-classification system has
+encouraged bad practices, such as adding non-relevant information. It should be
+modified to use a class hierarchy of generic errors. These errors should help
+classify the error, e.g. have UserError, PlatformError instead of logging an
+error type.
+
+### Progress Indicators
+
+`[Local]` Replace logs with progress indicators entirely. Give the user the
+option to choose between progress indicators and logs for local jobs.
+
+We should think about exporting progress tracking as a unified module, because
```

Comment:
tgdm supports progress trackers for multiple steps. 

I think we should re through their docs to get an idea how to handle dags.

Reply:
You can do some cools stuff with it. I did it with loops, but you can also control each progress bar separately.

https://github.com/vmware/versatile-data-kit/assets/91800778/dee5956a-7107-4ce2-88c8-b31662f77f52

https://tqdm.github.io/docs/tqdm/

```
>>> t = tqdm(total=filesize) # Initialise
>>> for current_buffer in stream:
...    ...
...    t.update(len(current_buffer))
>>> t.close()
```

Reply:
We can handle dags by having a progress bar for each job and a "parent" bar for the whole dag.